### PR TITLE
fix(providers): Buffer thinking across stream chunks for DeepSeek/Kimi tool calls

### PR DESCRIPTION
**Pull request by thepetk**
## Summary

The PR addresses the issue reported by https://github.com/aaif-goose/goose/issues/9676 for https://github.com/aaif-goose/goose/issues/9675. There's an issue with AI providers like `DeepSeek` or `Kimi` where some of their thinking models reason through a problem before answering. That said, on streaming responses, these providers send the thinking content in early chunks, then send the actual tool call in a later chunk (or the last one).

> See kimi docs: https://platform.kimi.ai/docs/guide/use-kimi-k2-thinking-model

This ends up in conversations involving a tool call; the next message sent to the API would fail with HTTP 400. The error was:

> reasoning_content in the thinking mode must be passed back to the API

This happens because `goose` was only looking for thinking content in the final chunk (i.e the tool-call chunk). By then, the thinking content had already been processed and discarded by an earlier stage of the stream. So the stored assistant message had the tool call, but there was no reasoning attached.

#### What changed

In `agent.rs`, I'm adding a `reasoning_buffer` that collects thinking content from every response chunk during a turn, not just the last one. When the tool-call chunk finally arrives, we drain that buffer and attach the buffered reasoning to the stored message, so the full context is preserved for the next API call.

Note: I'm also introducing a `reasoning_buffer_persisted` flag to avoid creating a duplicate reasoning message in the edge case where the thinking content was already saved via a different code path.

### Testing

Adding new tests in `crates/goose/tests/agent.rs` to cover:
- Reasoning split across multiple chunks before a tool call (that's the issue we are addressing here)
- Reasoning arriving together with the tool call (existing thinking behaviour to make sure we will avoid regressions)
- Non-thinking providers (i.e. classic `openai`); buffer stays empty, and there's no change in this behaviour.

### Related Issues
Relates to #9676 #9675
Discussion: n/a

### Screenshots/Demos (for UX changes)
Before: I was able to reproduce the issue with `moonshot` and `deepseek`

```
MOONSHOT_API_KEY=<api-key>
./target/debug/goose run \
--provider moonshot \
--model kimi-k2.6 \
--text "List the files in the current directory and tell me how many there are."

...

Ran into this error: Request failed: Bad request (400): thinking is enabled
but reasoning_content is missing in assistant tool call message at index 3.

Please retry if you think this is a transient or recoverable error.
```

After:  I was able to verify the fix works both with `moonshot` and `deepseek`.

```
MOONSHOT_API_KEY=<api-key>
./target/debug/goose run \
--provider moonshot \
--model kimi-k2.6 \
--text "List the files in the current directory and tell me how many there are."

...

There are **57 files and directories** ...
```

**Assisted-by**: Claude Sonnet 4.6

### CODE REVIEW

**Comment by thepetk** [comment]
Addressed feedback from the AI review and pushed a clean version of my branch

**Comment by thepetk** [comment]
Addressed the issue raised about `reasoning_buffer_persisted_idx` only tracking the first no-tool message with reasoning. Switched to a `Vec<usize>` so now all no-tool messages carrying Thinking are tracked. On the tool-call cases, `Thinking` is stripped from each of them before the full buffered reasoning is attached to the tool-call message, just to avoid duplication.

**Comment by jamadeo** [comment]
Thanks @thepetk ! Is it possible for this to live in the request formatter for the relevant providers? That seems a better place for it, so the provider can always produce a valid request rather than doing so on the conversation.

**Comment by thepetk** [comment]
> Thanks @thepetk ! Is it possible for this to live in the request formatter for the relevant providers? That seems a better place for it, so the provider can always produce a valid request rather than doing so on the conversation.

Thanks for the input @jamadeo. I think is a good point and tbh I like your approach more. Initially I thought too it could be an approach but I decided to work on the agent level as we already had this: https://github.com/aaif-goose/goose/blob/main/crates/goose/src/agents/agent.rs#L2174-L2188 & https://github.com/aaif-goose/goose/blob/main/crates/goose/src/agents/agent.rs#L2201-L2205.

I tried to move everything to the formatter. Let me know

**Comment by thepetk** [comment]
So let me add a summary here. As discussed above after @jamadeo's suggestion, I moved the reasoning preservation logic out of `agent.rs` and into the `openai` formatter, where it handles things transparently at request time without touching stored messages.

Basically 2 buffers were added to `format_messages_with_options`:
* one captures thinking-only messages and merges them into the next real assistant message
* second that propagates reasoning across split tool-call messages so they can be correctly reunited before the request is sent.

Additionally, I've addressed two a few issues raised by copilot. I've added unit tests in my fixes so to confirm they are resolved.

File: crates/goose-providers/src/formats/openai.rs [hunk 1]
```diff
@@ -182,12 +182,37 @@ pub fn format_messages_with_options(
...
+                saw_tool_response = false;
+            }
+        }
+
+        // A new assistant message after tool results creates a new turn
+        // To prevent reasoning from the previous turn leaking into the new one.
+        if options.preserve_thinking_context && message.role == Role::Assistant && saw_tool_response
+        {
+            tool_call_turn_reasoning.clear();
+            saw_tool_response = false;
```
[the remark this answers is not in the document]
Comment (thepetk) [reply crates/goose-providers/src/formats/openai.rs:213]: I think this is already covered by: `test_format_messages_carries_reasoning_to_all_split_tool_calls`

**Comment by thepetk** [comment]
Checked all comments from copilot and made sure we are covering the points with unit tests. Had to add one more here: [e38e3fe](https://github.com/aaif-goose/goose/pull/9704/commits/e38e3fed28a7181e80f9c04a9fcde29c969e66be)

**Review by jamadeo — APPROVED** [review]
thanks @thepetk ! I think this is the right place for it.
